src/cpp/rtps/transport/TCPTransportInterface.cpp

```diff
@@ -757,10 +757,44 @@ bool TCPTransportInterface::OpenOutputChannel(
             listening_port = config->listening_ports.front();
         }
 
+        bool local_lower_interface = false;
+        if (IPLocator::getPhysicalPort(physical_locator) == listening_port)
+        {
+            std::vector<Locator> list;
+            std::vector<fastrtps::rtps::IPFinder::info_IP> locNames;
+            get_ips(locNames);
+            for (const auto& infoIP : locNames)
+            {
+                Locator newloc(physical_locator);
+                if (transport_kind_ == LOCATOR_KIND_TCPv4)
+                {
+                    IPLocator::setIPv4(newloc, infoIP.locator);
+                }
+                else
+                {
+                    IPLocator::setIPv6(newloc, infoIP.locator);
+                }
+
+                if (is_interface_allowed(newloc))
+                {
+                    list.push_back(newloc);
+                }
+            }
+            if (list.empty())
+            {
+                EPROSIMA_LOG_ERROR(RTCP, "Could not find a valid local interface to connect to " << physical_locator);
+                return false;
+            }
+            else if (list.front() < physical_locator)
```

Comment:
I think the `empty()` condition could only happen if the only allowed interface is the loopback.
If this is the case, it would be impossible to discover a participant with the same listening port.
So I suggest to just check both conditions at the same time:

```suggestion
            if (!list.empty() && (list.front() < physical_locator))
```